File: Design and programming environments/Iteration 1/TicketManager/src/main/java/Repository/TicketRepository.java
package Repository;

import Domain.Event;
import Domain.Ticket;
import Utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;


import java.nio.DoubleBuffer;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class TicketRepository implements IRepository<Integer, Ticket> {

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public TicketRepository(Properties props){
        logger.info("Initializing TicketRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as [SIZE] from Tickets")){
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB Ticket Size " + e);
        }
        return 0;
    }

    @Override
    public void save(Ticket entity) {
        logger.traceEntry("saving ticket{}",entity);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Tickets values (?,?,?,?)")) {
            preStmt.setInt(1,entity.getId());
            preStmt.setString(2,entity.getBuyer());
            preStmt.setDouble(3,entity.getPrice());
            preStmt.setInt(4,entity.getEvent());
            int result = preStmt.executeUpdate();
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB Ticket Save "+e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting ticket with {}",integer);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from Tickets where id=?")){
            preStmt.setInt(1,integer);
            int result = preStmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.out.println("Error DB Ticket Delete "+e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Integer integer, Ticket entity) {
        logger.traceEntry("updating ticket with {}",integer);
        Connection con = dbUtils.getConnection();
        Ticket ticket = findOne(integer);
        if(ticket==null){
            return;
        }
        Integer newId=entity.getId();
        Integer newEvent=entity.getEvent();
        String newBuyer=entity.getBuyer();
        Double newPrice=entity.getPrice();

        try(PreparedStatement preStmt=con.prepareStatement("update Tickets"+
                " set id = ?,event = ?,buyer = ?,price = ? "+
                " where id = ? ")){
            preStmt.setInt(5,integer);
            preStmt.setInt(1,newId);
            preStmt.setInt(2,newEvent);
            preStmt.setString(3,newBuyer);
            preStmt.setDouble(4,newPrice);
            int result=preStmt.executeUpdate();
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Ticket Update " + e);
        }
        logger.traceExit();
    }

    @Override
    public Ticket findOne(Integer integer) {
        logger.traceEntry("finding ticket with id {}",integer);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select * from Tickets where id=?")){
            preStmt.setInt(1,integer);
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    int id= result.getInt("id");
                    int event = result.getInt("event");
                    String buyer=result.getString("buyer");
                    double price=result.getDouble("price");
                    Ticket ticket = new Ticket(id,event,buyer,price);
                    logger.traceExit(ticket);
                    return ticket;
                }
            }
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Ticket FindOne " + e);
        }
        logger.traceExit("No ticket found with id{}",integer);
        return null;
    }

    @Override
    public Iterable<Ticket> findAll() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Ticket> tickets=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Tickets")){
            try(ResultSet result=preStmt.executeQuery()){
                while(result.next()){
                    int id= result.getInt("id");
                    int event = result.getInt("event");
                    String buyer=result.getString("buyer");
                    double price=result.getDouble("price");
                    Ticket ticket = new Ticket(id,event,buyer,price);
                    tickets.add(ticket);
                }
            }
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Ticket FindAll " + e);
        }
        logger.traceExit(tickets);
        return tickets;
    }
}

File: Design and Programming Environments/Iteration 2/TicketManagerCSharp/TicketManagerCSharp/repository/TicketRepository.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using TicketManagerCSharp.domain;
using TicketManagerCSharp.utils;

namespace TicketManagerCSharp.repository
{
    class TicketRepository : IRepository<int, Ticket>
    {
        private static readonly ILog log = LogManager.GetLogger("EventRepository");

        IDictionary<String, string> props;

        public TicketRepository(IDictionary<String, string> props)
        {
            log.Info("Creating TicketRepository...");
            this.props = props;
        }

        public void Delete(int id)
        {
            log.InfoFormat("Entering Delete with new value {0}...", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from Tickets where id=@id";
                IDbDataParameter idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                comm.Parameters.Add(idParam);
                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't delete Ticket!");
                }
                log.InfoFormat("Exiting Delete");
            }
        }

        public IEnumerable<Ticket> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering findAll...");
            IList<Ticket> tickets = new List<Ticket>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,buyer,price,event from Users";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idTicket = dataR.GetInt32(0);
                        int idEvent = dataR.GetInt32(3);
                        string buyer = dataR.GetString(1);
                        double price = dataR.GetDouble(2);
                        Ticket ticket = new Ticket(idTicket, idEvent, buyer, price);
                        tickets.Add(ticket);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return tickets;
        }

        public Ticket FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,buyer,price,event, from Users where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idTicket = dataR.GetInt32(0);
                        int idEvent = dataR.GetInt32(3);
                        string buyer = dataR.GetString(1);
                        double price = dataR.GetDouble(2);
                        Ticket ticket = new Ticket(idTicket, idEvent, buyer, price);
                        log.InfoFormat("Exiting findOne with value{0}", ticket);
                        return ticket;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void Save(Ticket entity)
        {
            log.InfoFormat("Entering Save with new value {0}...", entity);
            var con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Tickets(buyer,price,event) values (@buyer,@price,@event)";
                var buyerParam = comm.CreateParameter();
                buyerParam.ParameterName = "@buyer";
                buyerParam.Value = entity.Buyer;
                comm.Parameters.Add(buyerParam);

                var priceParam = comm.CreateParameter();
                priceParam.ParameterName = "@price";
                priceParam.Value = entity.Price;
                comm.Parameters.Add(priceParam);

                var idEventParam = comm.CreateParameter();
                idEventParam.ParameterName = "@event";
                idEventParam.Value = entity.IdEvent;
                comm.Parameters.Add(idEventParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't add ticket!");
                }
                log.InfoFormat("Exiting Save");
            }
        }

        public int Size()
        {
            log.InfoFormat("Calculating Tickets database Size...");
            IDbConnection con = DBUtils.getConnection(props);
            int size = -1;
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from Tickets";

                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't calculate size!");
                }

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        size = dataR.GetInt32(0);
                    }
                }
            }
            log.InfoFormat("Finished calculation!");
            return size;
        }

        public void Update(int id, Ticket entity)
        {
            log.InfoFormat("Entering Update with value {0}", id);
            var con = DBUtils.getConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update Userss set buyer=@buyer,price=@price,event=@event where id=@id";
                var idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = entity.Id;
                comm.Parameters.Add(idParam);

                var buyerParam = comm.CreateParameter();
                buyerParam.ParameterName = "@buyer";
                buyerParam.Value = entity.Buyer;
                comm.Parameters.Add(buyerParam);

                var priceParam = comm.CreateParameter();
                priceParam.ParameterName = "@price";
                priceParam.Value = entity.Price;
                comm.Parameters.Add(priceParam);

                var idEventParam = comm.CreateParameter();
                idEventParam.ParameterName = "@event";
                idEventParam.Value = entity.IdEvent;
                comm.Parameters.Add(idEventParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't update Ticket!");
                }
                log.InfoFormat("Exiting Update");
            }
        }
    }
}
